### PR TITLE
Add CI test for local checkpointing

### PR DESCRIPTION
**Pull request by ananthsub**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds single node 2gpu test for local checkpointing support

**Collection**: llm

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: tests/functional_tests/L2_NeMo_2_llama3_local_ckpt.sh [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+mkdir -p /tmp/llm_tests/llama_local_ckpt_results \
+coverage run -a --data-file=/workspace/.coverage --source=/workspace/nemo tests/collections/llm/test_local_ckpt.py \
+    --devices=2 \
+    --log-dir=/tmp/llm_tests/llama_local_ckpt_results \
+    --async-save
```
Comment (ananthsub) [inline tests/functional_tests/L2_NeMo_2_llama3_local_ckpt.sh:18]:
```suggestion
mkdir -p /tmp/llm_tests/llama_local_ckpt_results

coverage run -a --data-file=/workspace/.coverage --source=/workspace/nemo tests/collections/llm/test_local_ckpt.py \
    --devices=2 \
    --log-dir=/tmp/llm_tests/llama_local_ckpt_results \
    --async-save
```

File: nemo/collections/llm/recipes/log/default.py [hunk 2]
```diff
@@ -86,22 +87,50 @@ def get_global_step_from_global_checkpoint_path(path: Union[str, Path]) -> int:
...
     # Format from above is {model_name}--{val_loss:.2f}-{step}-{consumed_samples}
-    assert "--" in dir_name, "Unexpected path format found for {path}"
+    assert "--" in dir_name, f"Unexpected path format found for {path}"
 
     # Find the parts after '--'
     double_dash_index = dir_name.index('--')
     remaining = dir_name[double_dash_index + 2 :]  # Skip the '--' itself
     parts = remaining.split('-')
-    assert len(parts) == 3, "Unexpected path format found for {path}"
+    assert len(parts) > 1, f"Unexpected path format found for {path}"
```
Comment (maanug-nv) [inline nemo/collections/llm/recipes/log/default.py:112]: I'm not sure why we require a specific checkpoint name format for this function, when that is supposed to be something the user can modify. But, does this line change affect that requirement?

Comment (ananthsub) [reply nemo/collections/llm/recipes/log/default.py:112]: this function is not required to use. it is offered as a convenience utility to match the default name format specified in L60 above. 

however, if users customize the checkpoint naming convention, they must provide their own function that gets the global step from the checkpoint name

File: nemo/collections/llm/recipes/log/default.py [hunk 3]
```diff
@@ -86,22 +87,50 @@ def get_global_step_from_global_checkpoint_path(path: Union[str, Path]) -> int:
...
     # Global step should be at index 1
-    step = int(parts[1])
+    step_part = parts[1]
+
+    if "=" in step_part:
+        # Handle "step=value" format (if metric name is auto-inserted)
+        step_value = step_part.split("=")
+        assert len(step_value) == 2, f"Expected 'step=<integer>' but found value '{step_value}'. Full path: {path}"
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply nemo/collections/llm/recipes/log/default.py:119]: as in this?

```suggestion
        assert len(step_value) == 2, f"Expected 'step=<integer>'. Full path: {path}"
```

File: tests/collections/llm/test_local_ckpt.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+        assert (
+            Path(args.log_dir).exists() or args.cleanup_log_dir
+        ), f"Log directory {args.log_dir} does not exist and cleanup was not requested."
+
+    # Log Directory Handling (Rank 0 Only)
+    log_dir_path = Path(args.log_dir)
+    if args.cleanup_log_dir and global_rank == 0:
+        if log_dir_path.exists():
+            # Use logger
+            logger.info(f"[Rank 0] Cleaning up log directory: {log_dir_path}")
```
Comment (maanug-nv) [inline tests/collections/llm/test_local_ckpt.py:236]: why is this necessary?

**Comment by maanug-nv** [comment]
Would be good to document in a bit more detail what behavior is being confirme , since the script is pretty long and would take another eng a while to parse. In top-level docstring, or in more comments throughout

File: nemo/collections/llm/recipes/log/default.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (maanug-nv) [reply nemo/collections/llm/recipes/log/default.py:119]: I forgot suggestions can be done in github. This is all i meant
```suggestion
        assert len(step_value) == 2, f"Expected 'step=<integer>' but found '{step_value}'. Full path: {path}"
```

File: tests/collections/llm/test_local_ckpt.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+            model=model_run1,
+            data=data,
+            trainer=trainer_run1,
+            log=nemo_logger_plugin,
+            optim=optim_run1,
+            resume=resume_logic,
+            tokenizer="data",
+        )
+
+        # If crash_step was 0 or None, training should complete Run 1 successfully
```
Comment (maanug-nv) [inline tests/collections/llm/test_local_ckpt.py:320]: is it possible for crash_step to be 0? based on asserts at lines 211-215, I don't think so?

File: tests/collections/llm/test_local_ckpt.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+from nemo.lightning.pytorch.optim.megatron import MegatronOptimizerModule
+from nemo.utils.import_utils import safe_import
+
+logger = logging.getLogger(__name__)
+
+res_module, HAVE_RES = safe_import("nvidia_resiliency_ext.ptl_resiliency")
+
+
+@dataclass
+class Llama3Config145M(Llama3Config):
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply tests/collections/llm/test_local_ckpt.py:73]: for simplicity here i want to avoid the nemo run / fdl dependency to build the config

File: .github/workflows/cicd-main-nemo2.yml [hunk 7]
```diff
@@ -47,6 +47,8 @@ jobs:
             runner: self-hosted-azure
           - script: L2_NeMo_2_llama3_straggler_detection
             runner: self-hosted-azure
+          - script: L2_NeMo_2_llama3_llama3_local_ckpt
```
Comment (ananthsub) [inline .github/workflows/cicd-main-nemo2.yml:50]:
```suggestion
          - script: L2_NeMo_2_llama3_local_ckpt
```

**Review by ko3n1g — APPROVED** [review]
good from automation perspective, let me know when to merge

**Comment by ananthsub** [comment]
@ko3n1g this is good to merge
